lib/input/blob_storage.go

```diff
@@ -0,0 +1,449 @@
+		if strings.Contains(conf.StorageConnectionString, "UseDevelopmentStorage=true;") {
+			client, err = storage.NewEmulatorClient()
+		} else {
+			client, err = storage.NewClientFromConnectionString(conf.StorageConnectionString)
+		}
+	} else {
+		client, err = storage.NewBasicClient(conf.StorageAccount, conf.StorageAccessKey)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("invalid azure storage account credentials: %v", err)
+	}
+
+	var objectScannerCtor codec.ReaderConstructor
+	if objectScannerCtor, err = codec.GetReader(conf.Codec, codec.NewReaderConfig()); err != nil {
+		return nil, fmt.Errorf("invalid azure storage codec: %v", err)
+	}
+
+	blobService := client.GetBlobService()
+	a := &azureBlobStorage{
+		conf:              conf,
+		objectScannerCtor: objectScannerCtor,
+		log:               log,
+		stats:             stats,
+		container:         blobService.GetContainerReference(conf.Container),
+		timeout:           timeout,
+	}
+
+	return a, nil
+}
+
+// ConnectWithContext attempts to establish a connection to the target Azure
+// Blob Storage container.
+func (a *azureBlobStorage) ConnectWithContext(ctx context.Context) error {
+	var err error
+	a.keyReader, err = newTargetReader(ctx, a.conf, a.log, a.container, a.timeout)
+	return err
+}
+
+func (a *azureBlobStorage) getObjectTarget(ctx context.Context) (*azurePendingObject, error) {
+	if a.object != nil {
+		return a.object, nil
+	}
+
+	target, err := a.keyReader.Pop(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	blobReference := a.container.GetBlobReference(target.key)
+	exists, err := blobReference.Exists()
+	if err != nil {
+		target.ackFn(ctx, err)
+		return nil, fmt.Errorf("failed to get blob reference: %v", err)
+	}
+
+	if !exists {
+		target.ackFn(ctx, err)
+		return nil, errors.New("blob does not exist")
+	}
+
+	obj, err := blobReference.Get(&storage.GetBlobOptions{Timeout: uint(a.timeout.Seconds())})
+	if err != nil {
+		target.ackFn(ctx, err)
+		return nil, err
+	}
+
+	object := &azurePendingObject{
+		target: target,
+		obj:    blobReference,
+	}
+	if object.scanner, err = a.objectScannerCtor(target.key, obj, target.ackFn); err != nil {
+		target.ackFn(ctx, err)
+		return nil, err
+	}
+
+	a.object = object
+	return object, nil
+}
+
+func blobStorageMsgFromPart(p *azurePendingObject, part types.Part) types.Message {
+	msg := message.New(nil)
+	msg.Append(part)
+
+	meta := msg.Get(0).Metadata()
+
+	meta.Set("blob_storage_key", p.target.key)
+	if p.obj.Container != nil {
+		meta.Set("blob_storage_container", p.obj.Container.Name)
+	}
+	meta.Set("blob_storage_last_modified", time.Time(p.obj.Properties.LastModified).Format(time.RFC3339))
+	meta.Set("blob_storage_last_modified_unix", strconv.FormatInt(time.Time(p.obj.Properties.LastModified).Unix(), 10))
+	meta.Set("blob_storage_content_type", p.obj.Properties.ContentType)
+	meta.Set("blob_storage_content_encoding", p.obj.Properties.ContentEncoding)
+
+	for k, v := range p.obj.Metadata {
+		meta.Set(k, v)
```

Comment:
I am a bit concerned that a user might define custom metadata with key names identical to the ones we used above, in which case we'd override them. I saw the same pattern [here](https://github.com/Jeffail/benthos/blob/2955a506cc48cf2afafdddea4eac4eb8ea8e9d5d/lib/input/aws_s3.go#L669), so I copied it over, but I wonder if it might be worth namespacing these somehow.

Reply:
I think this is probably fine, it means in a world where someone goes from blob storage -> something else -> blob storage -> something else again then the metadata would (depending on the "something elses") match the very first blob storage, which I think is probably a more useful default.

If in future this becomes a major problem for someone we can always add an optional `metadata_prefix` config field that would rename these keys, but as a user I would expect the default to be that my pre-existing metadata keys are unchanged.